portal_pages/tests/test_user_forms.py

```diff
@@ -0,0 +1,137 @@
+            instance=MarketplaceIndexPage(
+                title="Marketplace",
+                slug="marketplace",
+                live=True,
+            ))
+        self.marketplace_submission_url = self.marketplace_index_page.url + self.marketplace_index_page.reverse_subpage('submit')
+
+        self.marketplace_entry_form_data = { 
+            'date_start': '2015-01-01',
+            'title': 'new marketplace',
+            'biography': '<p>bio</p>',
+            # spambuster/honeypot fields
+            'email_field_email': '',
+            'company_company_email': 'nine'
+        }
+
+        # Add a case study index page under the home page
+        self.case_study_index_page = self.home_page.add_child(
+            instance=CaseStudyIndexPage(
+                title="Stories",
+                slug="stories",
+                live=True,
+            ))
+        self.case_study_submission_url = self.case_study_index_page.url + self.case_study_index_page.reverse_subpage('submit')
+
+        self.case_study_form_data = { 
+            'date': '2015-01-01',
+            'month_start': 12,
+            'year_start': 2014,
+            'title': 'new case study',
+            'summary': '<p>summary</p>',
+            # spambuster/honeypot fields
+            'email_field_email': '',
+            'company_company_email': 'nine'
+        }
+
+        # Add a blog index page under the home page
+        self.blog_index_page = self.home_page.add_child(
+            instance=BlogIndexPage(
+                title="Blog",
+                slug="blog",
+                live=True,
+            ))
+        self.blog_submission_url = self.blog_index_page.url + self.blog_index_page.reverse_subpage('submit')
+
+        self.blog_form_data = { 
+            'date': '2015-01-01',
+            'title': 'new blog',
+            'body': '<p>body</p>',
+            'submitter_email': 'fakeone@caktusgroup.com',
+            # spambuster/honeypot fields
+            'email_field_email': '',
+            'company_company_email': 'nine'
+        }
+
+    def test_marketplace_entry_form_valid(self):
+        # This should successfully add a single new marketplace entry
+        resp = self.client.post(self.marketplace_submission_url, self.marketplace_entry_form_data)
+        self.assertEqual(resp.status_code, 302)
+        self.assertEqual(resp['Location'], 'http://testserver/%s/submit-thank-you/' % self.marketplace_index_page.slug)
+        self.assertEqual(MarketplaceEntryPage.objects.all().count(), 1)
+
+    def test_marketplace_entry_form_invalid_honeypot(self):
+        # This should throw an ValidationError and not add the entry
+        self.marketplace_entry_form_data['company_company_email'] = 'Iamabot@bots.com'
+        resp = self.client.post(self.marketplace_submission_url, self.marketplace_entry_form_data)
+        self.assertEqual(resp.status_code, 200)
+        self.assertEqual(resp.context[0]['form'].errors['__all__'], ['Invalid submission'])
```

Comment:
This might become a test error instead of failure if for some reason errors does not have the `__all__` key. I would usually break this down into independent steps of checking that I got the form included in the response context, then got the number of errors expected, against the expected fields, with the expected messages:

```python
self.assertTrue('form' in resp.context)
form = resp.context['form']
self.assertEqual(len(form.errors), 1)
self.assertTrue('__all__' in form.errors)
self.assertEqual(form.errors['__all__'], ['Invalid submission'])
```